system/Validation/Validation.php

```diff
@@ -139,16 +139,20 @@ public function run(?array $data = null, ?string $group = null, ?string $dbGroup
             }
 
             $values = dot_array_search($field, $data);
-            $values = is_array($values) ? $values : [$values];
 
             if ($values === []) {
                 // We'll process the values right away if an empty array
                 $this->processRules($field, $setup['label'] ?? $field, $values, $rules, $data);
             }
 
-            foreach ($values as $value) {
-                // Otherwise, we'll let the loop do the job
-                $this->processRules($field, $setup['label'] ?? $field, $value, $rules, $data);
+            if (strpos($field, '*') !== false && is_array($values)) {
+                // Process multiple fields
+                foreach ($values as $value) {
+                    $this->processRules($field, $setup['label'] ?? $field, $value, $rules, $data);
+                }
+            } else {
+                // Process single field
+                $this->processRules($field, $setup['label'] ?? $field, $values, $rules, $data);
```

Comment:
Since this one is the same with L145, I think we can combine this with that?

Reply:
If I remove the `if` statement from L143, tests in `tests/system/Validaton/` passes.
But I'm not sure removing it causes no problem.

What input makes `$values` empty array?

Reply:
I added `continue` in the `if` to fix the bug that `$this->processRules()` run twice.

Reply:
> What input makes `$values` empty array?

This is raised in an issue report. I'll check.

Reply:
https://github.com/codeigniter4/CodeIgniter4/issues/4510#issuecomment-813532899

Reply:
I read the comment and the fix PR.
But I'm not sure we can remove the `if ($values === []) { ... }`.
I feel it is safer not to remove it.

Reply:
@paulbalandan I want to update `upgrade_416.rst`, so I merge this for now.